README.md

```diff
@@ -1 +1,252 @@
+   development mode.
+
+<table class="table table-condensed table-bordered table-params">
+  <colgroup>
+    <col class="col-param" />
+    <col class="param-description" />
+  </colgroup>
+  <thead>
+    <tr>
+      <th colspan="2">Attributes</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td><code>name</code></td>
+      <td>
+        <code>Name, required</code>
+        <p>A unique name for this rule.</p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>srcs</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          List of .java source files that will be compiled and passed on the
+          classpath to the GWT compiler.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>resources</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          List of resource files that will be passed on the classpath to the GWT
+          compiler, e.g. <code>.gwt.xml</code>, <code>.ui.xml</code>, and
+          <code>.css</code> files.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>modules</code></td>
+      <td>
+        <code>List of strings, required</code>
+        <p>
+          List of fully-qualified names of modules that will be passed to the GWT
+          compiler. Usually contains a single module name corresponding to the
+          application's <code>.gwt.xml</code> file.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>pubs</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          Files that will be copied directly to the output war, such as static
+          HTML or image resources. Not interpreted by the GWT compiler.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>deps</code></td>
+      <td>
+        <code>List of labels, optional</code>
+        <p>
+          List of other java_libraries on which the application depends. Both the
+          class jars and the source jars corresponding to each library as well as
+          their transitive dependencies will be passed to the GWT compiler's
+          classpath. These libraries may contain other <code>.gwt.xml</code>,
+          <code>.ui.xml</code>, etc. files as <code>resources</code>.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>output_root</code></td>
+      <td>
+        <code>String, optional</code>
+        <p>
+          Directory in the output war in which all outputs will be placed. By
+          default outputs are placed at the root of the war file.
+        </p>
+      </td>
+    </tr>
+    <tr>
+      <td><code>java_root</code></td>
+      <td>
+        <code>String, optional</code>
+        <p>
+          Directory relative to the workspace root that forms the root of the
+          Java package hierarchy (e.g. it contains a <code>com</code> directory).
+          By default this is `src/main/java`. If your Java files aren't under
```

Comment:
Yeah, this is my least favorite part of these rules, definitely open for suggestions on ways to improve it. GWT Dev mode is pretty weird and expects all of your source files to be on your classpath. If they aren't there, it isn't able to observe changes to them to trigger a recompilation when you refresh the browser, which mostly defeats the prupose.

The rules can tell which directory they're being run from relative to the workspace root, but I don't think they have enough information to be able to tell where the java package hierarchy starts (nothing to stop people from actually putting their classes under a `src.main.java` package). We could do more complex heuristics here like looking for a `com` directory and guessing that as the root, but I'd be worried about that being really confusing when the heuristic failed.

There is a check that ensures the directory that gets passed into dev mode actually exists and prints out a message about this flag if not, so I'm hoping that helps a bit with the discoverability. It's also not needed for normal compilation, just dev mode (which not everyone uses).